File: 508-MostFrequentSubtreeSum/508-MostFrequentSubtreeSum.java
/*

508. Most Frequent Subtree Sum
Medium

Given the root of a tree, you are asked to find the most frequent subtree sum. The subtree sum of a node is defined as the sum of all the node values formed by the subtree rooted at that node (including the node itself). So what is the most frequent subtree sum value? If there is a tie, return all the values with the highest frequency in any order.

Examples 1
Input:

  5
 /  \
2   -3

return [2, -3, 4], since all the values happen only once, return all of them in any order.

Examples 2
Input:

  5
 /  \
2   -5

return [2], since 2 happens twice, however -5 only occur once.

Note: You may assume the sum of values in any subtree is in the range of 32-bit signed integer.

*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    Map<Integer, Integer> lookupMap = new HashMap<>();
    int maxCount = 0;
    public int[] findFrequentTreeSum(TreeNode root) {
        dfs(root);
        List<Integer> res = new ArrayList<>();
        for (int s : lookupMap.keySet()) {
            if (lookupMap.get(s) == maxCount)
                res.add(s);
        }
        return res.stream().mapToInt(i -> i).toArray();
    }
    private int dfs(TreeNode node){
        if (node == null)
            return 0;
        int sum = dfs(node.left) + dfs(node.right) + node.val;
        lookupMap.put(sum, lookupMap.containsKey(sum) ? lookupMap.get(sum) + 1 : 1 );
        maxCount = Math.max(maxCount , lookupMap.get(sum));
        return sum;
        
    }
}

File: 508-MostFrequentSubtreeSum/508-MostFrequentSubtreeSum.py
"""

508. Most Frequent Subtree Sum

Given the root of a binary tree, return the most frequent subtree sum. If there is a tie, return all the values with the highest frequency in any order.

The subtree sum of a node is defined as the sum of all the node values formed by the subtree rooted at that node (including the node itself).

Example 1:

Input: root = [5,2,-3]
Output: [2,-3,4]

Example 2:

Input: root = [5,2,-5]
Output: [2]

Constraints:

    The number of nodes in the tree is in the range [1, 104].
    -105 <= Node.val <= 105

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findFrequentTreeSum(self, root: Optional[TreeNode]) -> List[int]:
        mapp = defaultdict(int)
        def dfs(node):
            if not node: return 0
            left = dfs(node.left)
            right = dfs(node.right)
            val = node.val + left + right
            mapp[val] += 1
            count[0] = max(count[0] , mapp[val])
            return val
        count  = [0]
        dfs(root)
        res = []
        for k, v in mapp.items():
            if v == count[0]:
                res.append(k)
        return res
